File: 250709/마을 구분하기/seperate-village.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const n = Number(input[0]);
const inputGrid = input.slice(1, n + 1).map(line => line.split(' ').map(Number));

// Please Write your code here.
const [dr, dc] = [[1, -1, 0, 0], [0, 0, 1, -1]];
const visitedGrid = Array.from({length: n}, () => new Array(n).fill(0));
const peopleArr = [];

function dfs(r, c) {
    let cnt = 1;
    visitedGrid[r][c] = 1;

    for (let i = 0; i < 4; i++) {
        const [nr, nc] = [r + dr[i], c + dc[i]];

        if (canGo(nr, nc)) {
            cnt += dfs(nr, nc);
        }
    }

    return cnt;
}

function inRange(r, c) {
    return r >= 0 && r < n && c >= 0 && c < n;
}

function searchGrid() {
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (canGo(i, j)) {
                peopleArr.push(dfs(i, j));
            }
        }
    }
}

function canGo(r, c) {
    // 1. 격자를 벗어나면 패스
    if (!inRange(r, c)) {
        return false;
    }
    // 2. 벽이 있으면 패스
    if (!inputGrid[r][c]) {
        return false;
    }
    // 3. 방문한 경우 패스
    if (visitedGrid[r][c]) {
        return false;
    }
    return true;
}

function makeAnswer() {
    searchGrid();
    peopleArr.sort((a, b) => a - b);

    console.log(peopleArr.length);
    peopleArr.forEach((elem) => {
        console.log(elem);
    })
}

makeAnswer();

File: 250719/마을 구분하기/seperate-village.py
n = int(input())
grid = [list(map(int, input().split())) for _ in range(n)]

# Please write your code here.
visited = [
    [False] * n
    for _ in range(n)
]

drs, dcs = [0, 0, 1, -1], [1, -1, 0, 0]

town_people_nums = []


def grid_search():
    for r in range(n):
        for c in range(n):
            if can_go(r, c):
                visited[r][c] = True
                town_people_nums.append(town_search(r, c))


def town_search(r, c):
    people_num = 1

    for dr, dc in zip(drs, dcs):
        nr, nc = r + dr, c + dc

        if (can_go(nr, nc)):
            visited[nr][nc] = True
            people_num += town_search(nr, nc)

    return people_num


def can_go(r, c):
    return in_range(r, c) and not visited[r][c] and grid[r][c]


def in_range(r, c):
    return r >= 0 and r < n and c >= 0 and c < n


def print_answer():
    grid_search()

    print(len(town_people_nums))

    town_people_nums.sort()
    for town_people_num in town_people_nums:
        print(town_people_num)


print_answer()